src/libstd/collections/hash/table.rs

```diff
@@ -655,7 +656,15 @@ impl<K, V, M> GapThenFull<K, V, M>
 fn calculate_layout<K, V>(capacity: usize) -> Result<(Layout, usize), LayoutErr> {
     let hashes = Layout::array::<HashUint>(capacity)?;
     let pairs = Layout::array::<(K, V)>(capacity)?;
-    hashes.extend(pairs)
+    hashes.extend(pairs).map(|(layout, _)| {
+        // LLVM seems to have trouble properly const-propagating pairs.align(),
+        // possibly due to the use of NonZeroUsize. This little hack allows it
+        // to generate optimal code.
```

Comment:
This is weird, the `NonZeroUsize::get(self)` function is marked `#[inline]` and the constructor is a `const fn`. Did you by chance happen to look at the generated LLVM-IR for `calculate_layout` ?

Reply:
Try the playground link that I posted above. If you remove the `map` then it generates worse code. What seems to happen is that LLVM will emit an overflowing multiply to calculate `pairs.size`, and will use the overflow flag to set `pairs.align` to either 0 (`LayoutErr`) or the actual alignment.

Reply:
Perhaps the comment should include the second sentence of that explanation?

Reply:
Every caller of `calculate_layout` only use one of the returned `Layout` or offset. Now that they’re mostly calculated separately, aren’t we better off having two separate functions?

Reply:
To me this feels like more of a temporary hack to deal with a LLVM bug, rather than a permanent solution. Logically the offset calculation belongs as part of the layout calculation.

Reply:
@nnethercote I added a reference to #51346 in the comment.